android-components/components/support/base/src/main/java/mozilla/components/support/base/android/NotificationsDelegate.kt

```diff
@@ -0,0 +1,174 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.support.base.android
+
+import android.Manifest.permission.POST_NOTIFICATIONS
+import android.annotation.SuppressLint
+import android.app.Notification
+import android.os.Build
+import android.os.RemoteException
+import androidx.activity.result.ActivityResultLauncher
+import androidx.activity.result.contract.ActivityResultContracts
+import androidx.annotation.RequiresApi
+import androidx.appcompat.app.AppCompatActivity
+import androidx.core.app.NotificationManagerCompat
+import androidx.lifecycle.Lifecycle
+
+typealias OnPermissionGranted = () -> Unit
+typealias OnPermissionRejected = () -> Unit
+
+/**
+ * Exception thrown when a [NotificationsDelegate] is not bound to any [AppCompatActivity]
+ */
+class UnboundHandlerException(message: String) : Exception(message)
+
+/**
+ * Handles showing notifications and asking permission, if needed.
+ *
+ * @param notificationManagerCompat a reference to [NotificationManagerCompat].
+ * @property activity The [AppCompatActivity] this handler is bound to.
+ * @property onPermissionGranted optional callback for handling permission acceptance.
+ * @property onPermissionRejected optional callback for handling permission refusal.
+ */
+class NotificationsDelegate(
+    val notificationManagerCompat: NotificationManagerCompat,
+) {
+    private val unboundHandlerException =
+        UnboundHandlerException("You must bind the NotificationPermissionHandler to an activity")
+    private var onPermissionGranted: OnPermissionGranted = { }
+    private var onPermissionRejected: OnPermissionRejected = { }
+    private val notificationPermissionHandler: MutableMap<AppCompatActivity, ActivityResultLauncher<String>> =
+        mutableMapOf()
+
+    /**
+     * Provides the context for a permission request.
+     */
+    @Suppress("unused")
+    fun bindToActivity(activity: AppCompatActivity) {
+        val activityResultLauncher =
+            activity.registerForActivityResult(ActivityResultContracts.RequestPermission()) { granted ->
+                if (granted) {
+                    onPermissionGranted.invoke()
+                } else {
+                    onPermissionRejected.invoke()
+                }
+            }
+
+        notificationPermissionHandler[activity] = activityResultLauncher
+    }
+
+    /**
+     * Removes activity reference from the [NotificationsDelegate]
+     */
+    @Suppress("unused")
+    fun unBindActivity(activity: AppCompatActivity) {
```

Comment:
Since we call this in `onDestroy`, I'm more inclined to having a weak ref wrapper since we know that `onDestroy` isn't always called.